src/bp/core/services/middleware/event-collector.ts

```diff
@@ -105,6 +164,24 @@ export class EventCollector {
     this.logger.info('Stopped')
   }
 
+  private buildQuery = (elements: BatchEvent[]) => {
+    const values = elements
+      .map(entry => {
+        const mappedValues = eventsFields.map(x => (x === 'event' ? JSON.stringify(entry[x]) : entry[x]) ?? null)
+        return this.knex.raw(`(${eventsFields.map(() => '?').join(',')})`, mappedValues).toQuery()
+      })
+      .join(',')
+
+    return this.knex
+      .raw(
+        `INSERT INTO ${this.TABLE_NAME}
+      (${eventsFields.map(x => `"${x}"`).join(',')}) VALUES ${values}
+        ON CONFLICT("id")
+        DO UPDATE SET event = EXCLUDED.event`
+      )
+      .toQuery()
+  }
```

Comment:
Not sure I get the fuss about this . can you add a jsdoc comment that explain why we need this in case someone else than you needs to work on this

Reply:
Each event is stored 3 times when debugging: on receive, in the middle and at completion. Since the batch runs each X seconds, it will either add the event to the batch, or update the existing one in place in the batch.

If the event was already inserted and needs to be updated, then it will update the existing entry instead of inserting it

Reply:
Gotcha. Nice